apisx/v1alpha1/listenerset_types.go

```diff
@@ -0,0 +1,582 @@
+// +kubebuilder:printcolumn:name="Programmed",type=string,JSONPath=`.status.conditions[?(@.type=="Programmed")].status`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// ListenerSet defines a set of additional listeners
+// to attach to an existing Gateway.
+type ListenerSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ListenerSet.
+	Spec ListenerSetSpec `json:"spec"`
+
+	// Status defines the current state of ListenerSet.
+	//
+	// +kubebuilder:default={conditions: {{type: "Accepted", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"},{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status ListenerSetStatus `json:"status,omitempty"`
+}
+
+// ListenerSetSpec defines the desired state of a ListenerSet.
+type ListenerSetSpec struct {
+	// ParentRef references the Gateway that the listeners are attached to.
+	ParentRef ParentGatewayReference `json:"parentRef,omitempty"`
+
+	// Listeners associated with this ListenerSet. Listeners define
+	// logical endpoints that are bound on this referenced parent Gateway's addresses.
+	//
+	// Listeners in a `Gateway` and their attached `ListenerSets` are concatenated
+	// as a list when programming the underlying infrastructure.
+	//
+	// Listeners should be merged using the following precedence:
+	//
+	// 1. "parent" Gateway
+	// 2. ListenerSet ordered by creation time (oldest first)
+	// 3. ListenerSet ordered alphabetically by “{namespace}/{name}”.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=64
+	// +kubebuilder:validation:XValidation:message="tls must not be specified for protocols ['HTTP', 'TCP', 'UDP']",rule="self.all(l, l.protocol in ['HTTP', 'TCP', 'UDP'] ? !has(l.tls) : true)"
+	// +kubebuilder:validation:XValidation:message="tls mode must be Terminate for protocol HTTPS",rule="self.all(l, (l.protocol == 'HTTPS' && has(l.tls)) ? (l.tls.mode == '' || l.tls.mode == 'Terminate') : true)"
+	// +kubebuilder:validation:XValidation:message="hostname must not be specified for protocols ['TCP', 'UDP']",rule="self.all(l, l.protocol in ['TCP', 'UDP']  ? (!has(l.hostname) || l.hostname == '') : true)"
+	// +kubebuilder:validation:XValidation:message="Listener name must be unique within the Gateway",rule="self.all(l1, self.exists_one(l2, l1.name == l2.name))"
+	//
+	// # TODO: Figure out how to allow empty port here
+	// #+kubebuilder:validation:XValidation:message="Combination of port, protocol and hostname must be unique for each listener",rule="self.all(l1, self.exists_one(l2, l1.port == l2.port && l1.protocol == l2.protocol && (has(l1.hostname) && has(l2.hostname) ? l1.hostname == l2.hostname : !has(l1.hostname) && !has(l2.hostname))))"
```

Comment:
note this TODO

Reply:
Ok this should become

```cel
self.all(l1, 
  !has(l1.port) || // Port is missing no conflict can ever occur
  self.exists_one(l2,
    has(l2.port) && // If L2 has no port than there can't be a conflict
   // Regular check since both listeners have a port
    l1.port == l2.port &&
    l1.protocol == l2.protocol  &&
    (has(l1.hostname) && has(l2.hostname) ? l1.hostname == l2.hostname : !has(l1.hostname) && !has(l2.hostname))
  )
)
```